designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala

```diff
@@ -0,0 +1,129 @@
+package pl.touk.nussknacker.ui.validation
+
+import cats.data.{Validated, ValidatedNel}
+import cats.implicits._
+import pl.touk.nussknacker.engine.api.NodeId
+import pl.touk.nussknacker.engine.api.context.PartSubGraphCompilationError
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError.MismatchParameter
+import pl.touk.nussknacker.engine.api.definition.MandatoryParameterValidator
+import pl.touk.nussknacker.engine.api.deployment.CustomActionCommand
+import pl.touk.nussknacker.engine.api.parameter.ParameterName
+import pl.touk.nussknacker.engine.deployment.{CustomActionDefinition, CustomActionParameter}
+import pl.touk.nussknacker.engine.graph.expression.Expression
+import pl.touk.nussknacker.restmodel.CustomActionRequest
+
+class CustomActionValidator(allowedAction: CustomActionDefinition) {
+
+  def validateCustomActionParams(
+      request: CustomActionRequest
+  ): ValidatedNel[PartSubGraphCompilationError, Unit] = {
+
+    implicit val nodeId: NodeId = NodeId(allowedAction.actionName.value)
+    val customActionParams      = allowedAction.parameters
+
+    val missingParams = checkForMissingParams(request.params, customActionParams)
+    val checkedParams = validateParams(request.params, customActionParams)
+
+    missingParams.combine(checkedParams)
+  }
+
+  private def checkForMissingParams(
+      requestParamsMap: Map[String, String],
+      customActionParams: List[CustomActionParameter]
+  )(implicit nodeId: NodeId): ValidatedNel[PartSubGraphCompilationError, Unit] = {
+
+    val paramsFromRequest = requestParamsMap.keys.toSet
+    val mandatoryParams = customActionParams.collect {
+      case param if param.validators.contains(MandatoryParameterValidator) && !paramsFromRequest.contains(param.name) =>
+        MandatoryParameterValidator.isValid(ParameterName(param.name), Expression.spel(""), None, None)
+    }
+    mandatoryParams match {
+      case Nil      => Validated.Valid(())
+      case nonEmpty => nonEmpty.traverse(_.toValidatedNel).map(_ => ())
+    }
+  }
+
+  private def validateParams(
+      requestParamsMap: Map[String, String],
+      customActionParams: List[CustomActionParameter]
+  )(implicit nodeId: NodeId): ValidatedNel[PartSubGraphCompilationError, Unit] = {
+
+    requestParamsMap match {
+      case paramsMap if paramsMap.nonEmpty =>
+        handleNonEmptyParamsRequest(paramsMap, customActionParams)
+      case emptyMap if emptyMap.isEmpty =>
```

Comment:
Why the empty map is some special case? Eventually you traverse all params so what's the difference?

Reply:
In case that the map is empty I'm checking if there are validators that will pop off for an empty request (like MandatoryParameter). Because of this i have divided this check into two cases.

Reply:
And for one element map with missing some parameter value that is mandatory you don't need to do the same?

Reply:
you are right, I forgot to clean this up when i extracted 
`private def checkForMissingParams(
      requestParamsMap: Map[String, String],
      customActionParams: List[CustomActionParameter]`
      This emptyParamsMapRequest can probably go along with the None expression check. I will update it.